[core] Fix append path from .spacemacs.env

mirubiri opened the pull request:
Fixes the issue described in #12403 and potentially #12396 #10906.

The function `load-env-vars-set-env` is appending to `exec-path` the PATH values found in `.spacemacs.env`. 

This means that the directories defined in `exec-path` will be searched first instead of the ones we have defined in `.spacemacs.env` PATH variable. 

Usually `exec-path` already includes `/bin;/usr/bin;/usr/local/bin`, so it's not possible to look for binaries in the directories specified in our custom PATH variable. This is needed when working with shims, which is something we usually found in tools like `asdf`,`rbenv` or `brew`..

This PR preprends the `.spacemacs.env` PATH variable to `exec-path` to avoid this issues.

smile13241324 reviewed (approved):
Strange, actually this should have appended at the end of `exec-path` already given the `append` argument. When I check the content of `exec-path` on my linux system it also seems to be in the right order.

Anyway your code is much more functional and if it fixes issues on other oses it should definitely be merged.

smile13241324 commented on core/libs/load-env-vars.el:
```diff
@@ -87,12 +87,10 @@
-              (setenv "PATH" (mapconcat convert-to-os-path exec-path
-                                        path-separator)))
+              (setq exec-path (append
+                               (mapcar (lambda (path) (subst-char-in-string ?\\ ?/ path)) paths)
```
Ups, just found a possible issue, the original code did also call `directory-file-name` on the paths. Please restore that part of the logic otherwise we may risk issues on certain systems.

mirubiri replied:
Upss, restored!

smile13241324 reviewed (changes requested):
Looks like I was too fast, I think I have found a possible issue, please check my comment.
As this is affecting the core some manual testing should happen on this PR before it is merged.

mirubiri commented:
> Strange, actually this should have appended at the end of exec-path already given the append argument. When I check the content of exec-path on my linux system it also seems to be in the right order.

It is appended at the end, but the problem is that `exec-path` might already contain paths for binaries like `usr/bin`. I think it's interesting to `prepend` the PATH var defined in `.spacemacs.env` to `exec-path` instead. 

For example I always want that the first section of my PATH to be `~/.asdf/shims` to look for binaries, but it doesn't have any effect if `exec-path` already includes `usr/bin`. If I'm allowed to prepend my PATH to `exec-path` problem solved and my configurations will take effect. This should be valuable for any system I guess.

mirubiri commented on core/core-env.el:
```diff
@@ -63,7 +63,7 @@ current contents of the file will be overwritten."
           "# duplicate settings for a variable, only the first setting is effective.\n"
           "# PATH is a special case: all PATH settings are read, each non-duplicate\n"
-          "# directory entry is appended to the `exec-path' variable, and then PATH is\n"
+          "# directory entry is prepended to the `exec-path' variable, and then PATH is\n"
```
@smile13241324 I have also updated the generated documentation here to reflect this change

mirubiri commented:
@smile13241324 Are you launching it from the console? These are my values when launching it in osx with emacs-mac port.

**Withouth this PR:**

Original: (/usr/local/bin /usr/bin /bin /usr/sbin /sbin /usr/local/Cellar/emacs-mac/emacs-27.1-mac-8.0/libexec/emacs/27.1/x86_64-apple-darwin19.6.0)

Modified: (/usr/local/bin /usr/bin /bin /usr/sbin /sbin /usr/local/Cellar/emacs-mac/emacs-27.1-mac-8.0/libexec/emacs/27.1/x86_64-apple-darwin19.6.0 /Users/alex/.asdf/shims /usr/local/opt/asdf/bin /Users/alex/bin /usr/local/sbin /Users/alex/Library/Python/3.7/bin)

**With this PR:**

Original: (/usr/local/bin /usr/bin /bin /usr/sbin /sbin /usr/local/Cellar/emacs-mac/emacs-27.1-mac-8.0/libexec/emacs/27.1/x86_64-apple-darwin19.6.0)

Modified: (/Users/alex/.asdf/shims /usr/local/opt/asdf/bin /usr/local/bin /usr/bin /bin /usr/sbin /sbin /Users/alex/bin /usr/local/bin /usr/local/sbin /Users/alex/Library/Python/3.7/bin /usr/local/bin /usr/bin /bin /usr/sbin /sbin /usr/local/Cellar/emacs-mac/emacs-27.1-mac-8.0/libexec/emacs/27.1/x86_64-apple-darwin19.6.0)

PATH=/Users/alex/.asdf/shims:/usr/local/opt/asdf/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/alex/bin:/usr/local/bin:/usr/local/sbin:/Users/alex/Library/Python/3.7/bin


As you can see, starting my path starts with `/Users/alex/.asdf/shims`, without this PR my PATH is appended to the end of exec-path and with this PR, my PATH is prepended to exec-path.

I'm going to remove make some changes to remove the duplicates

mirubiri commented:
Duplicates removed 🤞

smile13241324 commented:
Thanks 👍 , I'll have another go in the next view days.

smile13241324 commented:
Looks good now, and the code is still much clearer than before. 

The order of paths does not change for me though but this may be system related. I am not starting emacs from the console I am running it from latest Gnome on Manjaro. Anyway if it solves issues on one Os and does not hurt the others it is worth to be merged.
